File: android/src/main/java/com/reactlibrary/scene/nodes/UiScrollViewNode.kt
/*
 * Copyright (c) 2019 Magic Leap, Inc. All Rights Reserved
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.reactlibrary.scene.nodes

import android.content.Context
import android.os.Bundle
import android.os.Handler
import android.os.Looper
import android.view.LayoutInflater
import android.view.MotionEvent
import android.view.View
import com.facebook.react.bridge.ReadableMap
import com.google.ar.sceneform.Node
import com.google.ar.sceneform.math.Vector3
import com.reactlibrary.R
import com.reactlibrary.ar.ViewRenderableLoader
import com.reactlibrary.scene.nodes.base.TransformNode
import com.reactlibrary.scene.nodes.base.UiNode
import com.reactlibrary.scene.nodes.props.Bounding
import com.reactlibrary.scene.nodes.views.CustomScrollView
import com.reactlibrary.utils.PropertiesReader
import com.reactlibrary.utils.Utils.Companion.metersToPx
import com.reactlibrary.utils.Utils.Companion.pxToMeters
import com.reactlibrary.utils.Vector2
import com.reactlibrary.utils.putDefaultString
import kotlin.math.max
import kotlin.math.min

open class UiScrollViewNode(
        initProps: ReadableMap,
        context: Context,
        viewRenderableLoader: ViewRenderableLoader) :
        UiNode(initProps, context, viewRenderableLoader, useContentNodeAlignment = false) {

    companion object {
        // properties
        const val PROP_SCROLL_BOUNDS = "scrollBounds"
        const val PROP_SCROLL_DIRECTION = "scrollDirection"

        const val DEFAULT_WIDTH = 1.0F
        const val DEFAULT_HEIGHT = 1.0F
        const val DEFAULT_SCROLL_DIRECTION = "vertical"

        const val BAR_THICKNESS_RATIO = 0.03F // relative to min (width, height)
        const val BAR_MINIMUM_THICKNESS = 0.05F // in meters
        const val HIDDEN_BAR_THICKNESS = 0

        const val LAYOUT_LOOP_DELAY = 50L
        const val Z_ORDER_OFFSET = 1e-5F
    }

    var onScrollChangeListener: ((position: Vector2) -> Unit)? = null

    protected var onContentSizeChangedListener: ((contentSize: Vector2) -> Unit)? = null

    protected var vBarNode: UiScrollBarNode? = null
        private set

    protected var hBarNode: UiScrollBarNode? = null
        private set

    private var content: TransformNode? = null
    private var contentBounds = Bounding()

    private var requestedContentPosition = Vector2()
    private var looperHandler = Handler(Looper.getMainLooper())

    init {
        properties.putDefaultString(PROP_SCROLL_DIRECTION, DEFAULT_SCROLL_DIRECTION)
        layoutLoop()
    }

    // Function by which ViewWrapper delivers intercepted motion events.
    fun onTouchEvent(event: MotionEvent): Boolean {
        val viewBounds = getScrollBounds()
        event.setLocation(
                metersToPx(event.x - viewBounds.left, context).toFloat(),
                metersToPx(-event.y + viewBounds.top, context).toFloat())
        return view.onTouchEvent(event)
    }

    fun stopNestedScroll() {
        view.stopNestedScroll()
    }

    override fun provideView(context: Context): View {
        return LayoutInflater.from(context).inflate(R.layout.scroll_view, null)
    }

    override fun provideDesiredSize(): Vector2 {
        val propBounds = PropertiesReader.readAABB(properties, PROP_SCROLL_BOUNDS)
        return if (propBounds != null) {
            Vector2(propBounds.getWidth(), propBounds.getHeight())
        } else {
            Vector2(DEFAULT_WIDTH, DEFAULT_HEIGHT)
        }
    }

    override fun setupView() {
        super.setupView()

        if (vBarNode != null) {
            val thickness = calculateBarThickness(size)
            val thicknessPx = metersToPx(thickness, context)
            (view as CustomScrollView).vBar?.setThickness(thicknessPx)
        }

        if (hBarNode != null) {
            val thickness = calculateBarThickness(size)
            val thicknessPx = metersToPx(thickness, context)
            (view as CustomScrollView).hBar?.setThickness(thicknessPx)
        }

        val scrollView = view as CustomScrollView
        scrollView.onScrollChangeListener = { position: Vector2 ->
            update(position)
            this.onScrollChangeListener?.invoke(position)
        }
    }

    override fun applyProperties(props: Bundle) {
        super.applyProperties(props)

        if (props.containsKey(PROP_SCROLL_BOUNDS)) {
            setNeedsRebuild()
        }

        setScrollDirection(props)
    }

    override fun addContent(child: Node) {
        if (child is UiScrollBarNode) {
            addScrollBar(child)
            return
        }

        if (content != null) {
            return
        }
        super.addContent(child)

        if (child !is TransformNode) {
            return
        }
        this.content = child
    }

    override fun removeContent(child: Node) {
        super.removeContent(child)
        when (child) {
            content -> content = null
            hBarNode -> {
                (view as CustomScrollView).hBar?.setThickness(HIDDEN_BAR_THICKNESS)
                hBarNode = null
            }
            vBarNode -> {
                (view as CustomScrollView).vBar?.setThickness(HIDDEN_BAR_THICKNESS)
                vBarNode = null
            }
        }
    }

    override fun getContentPosition(): Vector2 {
        // When translating clip bounds to node's local coordinate
        // system we use localPositions. However due to ARCore nature
        // localPosition of ScrollViewNode direct descendant can change
        // randomly for some milliseconds after being written to. Thus
        // we can't reliably use it. To solve the problem we add child
        // localPosition to translation - the child node will subtract
        // it in it's setClipBounds method, effectively zeroing it.
        // Then, instead of zeroed child localPosition we use
        // requestedContentPosition.
        val position = content?.localPosition ?: Vector3()
        return Vector2(
                requestedContentPosition.x - position.x,
                requestedContentPosition.y - position.y
        )
    }

    override fun setAlignment(props: Bundle) {
        // Alignment cannot be changed for scroll view according to Lumin.
        // It is hardcoded to center-center
    }

    override fun onDestroy() {
        super.onDestroy()
        // stop the layout loop
        looperHandler.removeCallbacksAndMessages(null)
    }

    protected fun calculateBarThickness(containerSize: Vector2): Float {
        val parentBasedThickness = min(containerSize.x, containerSize.y) * BAR_THICKNESS_RATIO
        return max(parentBasedThickness, BAR_MINIMUM_THICKNESS)
    }

    private fun addScrollBar(scrollBarNode: UiScrollBarNode) {
        val thickness = calculateBarThickness(size)
        val thicknessPx = metersToPx(thickness, context)

        if (scrollBarNode.orientation == UiScrollBarNode.ORIENTATION_VERTICAL) {
            if (vBarNode == null) {
                super.addContent(scrollBarNode)
                vBarNode = scrollBarNode
                (view as CustomScrollView).vBar?.setThickness(thicknessPx)
            }
        } else {
            if (hBarNode == null) {
                super.addContent(scrollBarNode)
                hBarNode = scrollBarNode
                (view as CustomScrollView).hBar?.setThickness(thicknessPx)
            }
        }
    }

    private fun layoutLoop() {
        looperHandler.postDelayed({
            content?.let { it ->
                // check if content bounds changed
                val newContentBounds = it.getContentBounding()
                if (!Bounding.equalInexact(contentBounds, newContentBounds)) {
                    this.contentBounds = newContentBounds
                    onContentSizeChangedListener?.invoke(newContentBounds.size())
                }
                layout()
            }
            layoutLoop()
        }, LAYOUT_LOOP_DELAY)
    }

    private fun layout() {
        val scrollView = (view as CustomScrollView)
        val contentSize = contentBounds.size()
        scrollView.contentSize = Vector2(
                metersToPx(contentSize.x, context).toFloat(),
                metersToPx(contentSize.y, context).toFloat())
        update(scrollView.position)
    }

    private fun update(viewPosition: Vector2) {
        content?.let { content ->
            val contentBounds = content.getContentBounding()
            val viewBounds = getScrollBounds()
            val alignTopLeft = Vector2(
                    viewBounds.left - contentBounds.left,
                    viewBounds.top - contentBounds.top)

            val contentSize = contentBounds.size()
            val viewSize = viewBounds.size()
            val possibleTravel = (contentSize - viewSize).coerceAtLeast(0F)
            val travel = Vector2(
                    -possibleTravel.x * viewPosition.x,
                    possibleTravel.y * viewPosition.y)

            requestedContentPosition = alignTopLeft + travel

            val clipBounds = viewBounds.translate(-getContentPosition())
            content.setClipBounds(clipBounds)

            // Moving content up in z-plane, so it'll receive touch events first.
            content.localPosition = Vector3(
                    requestedContentPosition.x,
                    requestedContentPosition.y,
                    Z_ORDER_OFFSET)

        }
    }

    private fun getScrollBounds(): Bounding {
        val hBarHeightPx = (view as CustomScrollView).hBar?.height ?: 0
        val vBarWidthPx = (view as CustomScrollView).vBar?.width ?: 0

        val hBarHeightMeters = pxToMeters(hBarHeightPx, context)
        val vBarWidthPxMeters = pxToMeters(vBarWidthPx, context)

        return Bounding(
                -size.x / 2F,
                -size.y / 2F + hBarHeightMeters,
                size.x / 2F - vBarWidthPxMeters,
                size.y / 2F)
    }

    private fun setScrollDirection(props: Bundle) {
        val value = props.getString(PROP_SCROLL_DIRECTION)
        if (value != null) {
            (view as CustomScrollView).scrollDirection = value
        }
    }


}

File: ios/RNMagicScript/components/UiNodes/UiScrollViewNode.swift
//
//  Copyright (c) 2019 Magic Leap, Inc. All Rights Reserved
//
//  Licensed under the Apache License, Version 2.0 (the "License");
//  you may not use this file except in compliance with the License.
//  You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
//  Unless required by applicable law or agreed to in writing, software
//  distributed under the License is distributed on an "AS IS" BASIS,
//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  See the License for the specific language governing permissions and
//  limitations under the License.
//

import SceneKit

@objc open class UiScrollViewNode: UiNode {

    @objc override var alignment: Alignment {
        get { return .centerCenter }
        set { }
    }
    @objc var scrollDirection: ScrollDirection = .horizontal {
        didSet { setNeedsLayout() }
    }
    // Scroll speed in scene units per second.
    @objc var scrollSpeed: CGFloat = 0.1
    @objc var scrollOffset: SCNVector3 = SCNVector3Zero {
        didSet { setNeedsLayout() }
    }
    fileprivate var _scrollValue: CGFloat = 0
    @objc var scrollValue: CGFloat {
        get { return _scrollValue }
        set {
            let clampedValue: CGFloat = Math.clamp(newValue, 0.0, 1.0)
            if (_scrollValue != clampedValue) {
                _scrollValue = clampedValue
                scrollBar?.thumbPosition = clampedValue
                setNeedsLayout()
            }
        }
    }
    //@objc
    var scrollBounds: (min: SCNVector3, max: SCNVector3)? {
        didSet { invalidateClippingPlanes = true; setNeedsLayout() }
    }
    @objc var scrollBarVisibility: ScrollBarVisibility = .auto {
        didSet { updateScrollBarVisibility() }
    }

    @objc public var onScrollChanged: ((_ sender: UiNode, _ value: CGFloat) -> (Void))?

    fileprivate weak var scrollBar: UiScrollBarNode?
    fileprivate weak var scrollContent: TransformNode?
    fileprivate var proxyNode: SCNNode!
    fileprivate var invalidateClippingPlanes: Bool = false

    deinit {
        scrollContent?.resetClippingPlanes()
    }

    @objc override func setupNode() {
        super.setupNode()

        proxyNode = SCNNode()
        contentNode.addChildNode(proxyNode)
    }

    @objc override func hitTest(ray: Ray) -> TransformNode? {
        guard let _ = selfHitTest(ray: ray) else { return nil }

        let nodes: [TransformNode?] = [scrollContent, scrollBar]
        for node in nodes {
            if let hitNode = node?.hitTest(ray: ray) {
                return hitNode
            }
        }

        return self
    }

    @objc override func update(_ props: [String: Any]) {
        super.update(props)

        if let scrollDirection = Convert.toScrollDirection(props["scrollDirection"]) {
            self.scrollDirection = scrollDirection
        }

        if let scrollSpeed = Convert.toCGFloat(props["scrollSpeed"]) {
            self.scrollSpeed = scrollSpeed
        }

        if let scrollOffset = Convert.toVector3(props["scrollOffset"]) {
            self.scrollOffset = scrollOffset
        }

        if let scrollValue = Convert.toCGFloat(props["scrollValue"]) {
            self.scrollValue = scrollValue
        }

        if let scrollBounds = props["scrollBounds"] as? [String: Any],
            let min = Convert.toVector3(scrollBounds["min"]),
            let max = Convert.toVector3(scrollBounds["max"]) {
            self.scrollBounds = (min: min, max: max)
        }

        if let scrollBarVisibility = Convert.toScrollBarVisibility(props["scrollBarVisibility"]) {
            self.scrollBarVisibility = scrollBarVisibility
        }
    }

    @discardableResult
    @objc override func addChild(_ child: TransformNode) -> Bool {
        if child is UiScrollBarNode {
            guard scrollBar == nil else { return false }
            scrollBar = child as? UiScrollBarNode
            contentNode.addChildNode(child)
            setNeedsLayout()
            updateScrollBarVisibility()
            return true
        }

        guard scrollContent == nil else { return false }
        scrollContent = child
        proxyNode.addChildNode(child)
        invalidateClippingPlanes = true
        setNeedsLayout()
        return true
    }

    @objc override func removeChild(_ child: TransformNode) {
        if child == scrollBar {
            scrollBar?.removeFromParentNode()
            scrollBar = nil
            setNeedsLayout()
        } else if child == scrollContent {
            scrollContent?.removeFromParentNode()
            scrollContent = nil
            invalidateClippingPlanes = true
            setNeedsLayout()
        }
    }

    @objc override func _calculateSize() -> CGSize {
        guard let scrollBounds = scrollBounds else { return CGSize.zero }
        let width: CGFloat = CGFloat(scrollBounds.max.x - scrollBounds.min.x)
        let height: CGFloat = CGFloat(scrollBounds.max.y - scrollBounds.min.y)
        return CGSize(width: width, height: height)
    }

    @objc override func getBounds(parentSpace: Bool = false, scaled: Bool = true) -> CGRect {
        guard let scrollBounds = scrollBounds else { return CGRect.zero }
        let min = scrollBounds.min
        let size = getSize(scaled: scaled)
        let origin = CGPoint(x: CGFloat(min.x), y: CGFloat(min.y))
        let offset = parentSpace ? CGPoint(x: CGFloat(localPosition.x), y: CGFloat(localPosition.y)) : CGPoint.zero
        return CGRect(origin: origin + offset, size: size)
    }

    @objc override func updateLayout() {
        scrollBar?.layoutIfNeeded()
        scrollContent?.layoutIfNeeded()

        // Update scroll content
        let bounds = getBounds()
        let scrollSize = bounds.size
        let contentSize: CGSize
        if let scrollContent = scrollContent {
            contentSize = scrollContent.getSize()
            let contentBounds = scrollContent.getBounds(parentSpace: true)
            let contentOffset = SCNVector3(bounds.minX - contentBounds.minX, bounds.maxY - contentBounds.maxY, 0)
            proxyNode.position = contentOffset
        } else {
            contentSize = CGSize.zero
            proxyNode.position = SCNVector3Zero
        }

        var shift: CGPoint = CGPoint.zero
        switch scrollDirection {
        case .horizontal:
            shift.x = -scrollValue * max(0, contentSize.width - scrollSize.width)
        case .vertical:
            shift.y = scrollValue * max(0, contentSize.height - scrollSize.height)
        }
        proxyNode.position += SCNVector3(shift.x, shift.y, 0)

        // Update scroll bar
        scrollBar?.thumbPosition = scrollValue
    }

    @objc override func postUpdate() {
        guard let scrollBounds = scrollBounds else { return }

        // Update clipping planes
        if invalidateClippingPlanes {
            invalidateClippingPlanes = false

            let min = scrollBounds.min
            let max = scrollBounds.max
            let planes: [SCNVector4] = [
                SCNVector4( 1, 0, 0,-min.x),
                SCNVector4(-1, 0, 0, max.x),
                SCNVector4(0, 1, 0,-min.y),
                SCNVector4(0,-1, 0, max.y),
                SCNVector4(0, 0, 1,-min.z),
                SCNVector4(0, 0,-1, max.z),
            ]

            let worldSpacePlanes: [SCNVector4] = planes.map { convertPlane(Plane(vector: $0), to: nil).toVector4() }
            scrollContent?.setClippingPlanes(worldSpacePlanes)
        }
    }


    fileprivate func updateScrollBarVisibility() {
        switch scrollBarVisibility {
            case .always:
                scrollBar?.visible = true
            case .auto:
                scrollBar?.setVisible(true, animated: true, delay: 0.0) { [weak self] in
                    self?.scrollBar?.setVisible(false, animated: true, delay: 2.0)
                }
            case .off:
                scrollBar?.visible = false
        }
    }
}

extension UiScrollViewNode: Dragging {
    var dragAxis: Ray? {
        guard let scrollBounds = scrollBounds else { return nil }
        let min = scrollBounds.min
        let max = scrollBounds.max
        let center: SCNVector3 = 0.5 * (min + max)
        let direction: SCNVector3
        if scrollDirection == .horizontal {
            direction = SCNVector3(center.x - max.x, center.y, center.z)
        } else {
            direction = SCNVector3(center.x, max.y - center.y, center.z)
        }

        return Ray(begin: center - direction, direction: direction.normalized(), length: CGFloat(2 * direction.length()))
    }

    var dragRange: CGFloat {
        guard let contentSize = scrollContent?.getSize() else { return 0 }
        let size = getSize()
        return (scrollDirection == .horizontal) ? contentSize.width - size.width : contentSize.height - size.height
    }

    var dragValue: CGFloat {
        get { return scrollValue }
        set {
            let prevScrollValue = scrollValue
            scrollValue = newValue
            if prevScrollValue != scrollValue {
                layoutIfNeeded()
                onScrollChanged?(self, scrollValue)

                if scrollBarVisibility == .auto {
                    scrollBar?.setVisible(true, animated: true, delay: 0.0) { [weak self] in
                        self?.scrollBar?.setVisible(false, animated: true, delay: 2.0)
                    }
                }
            }
        }
    }
}
